design/one-pager-stacks-security-isolation.md

```diff
@@ -0,0 +1,231 @@
+
+This design doc proposes a new approach for the Crossplane Stack Manager (SM) that will support a smaller set of security permissions and increased control for the installation of stacks.
+
+## Design Summary
+
+The key differences in this design update are in the **scope** of the permissions that stacks can be granted within the control plane.
+
+* **Cluster Stacks** will be given permissions to watch, access, and create resources at a **cluster level**, i.e. across **all namespaces**.
+* **Stacks** will be given permissions only within a **single namespace**.
+* All stacks will be granted permissions to a small subset of core Kubernetes resources as well as access to any CRDs they own and depend on.
+
+This means that stacks should not be able to directly affect any resources outside of the specific namespace they are installed into, while cluster stacks operate at a higher level of privilege and can affect the entire cluster.
+This is a familiar concept to the Kubernetes permission scoping model of `ClusterRole` and `Role`.
+
+### CRD Scoping
+
+To avoid a common point of confusion, let's review the available scoping of CRDs.
+Custom resource **instances** can be scoped to either a single namespace or to the entire cluster.
+It's important to note that this refers to the instances, not the CRD type definitions.
+From the [Kubernetes documentation](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#create-a-customresourcedefinition):
+
+> CustomResourceDefinitions themselves are non-namespaced and are available to all namespaces
+
+Therefore, any time a CRD is created, it will also be at the cluster scope, even if instances of the CRD are scoped to a namespace.
+This has caused confusion in the past, so it's important to clear that up here.
+
+## Allowed Resource Access
+
+As previously stated, Crossplane is not intended to be a typical/full Kubernetes cluster to run arbitrary workloads.
+Once installed into the control plane, stacks will mostly target external resources and a known set of CRDs that they own and depend on.
+Therefore, we can lock down the allowed permissions for stacks to a subset of all available permissions in the Kubernetes API.
+
+The `rbac.yaml` file that allows a stack to specify the permissions it requires will be removed and no longer supported.
+Instead, the permissions for a stack can be considered **implicit**.
+The following permissions will be granted to a stack when it is installed:
+
+* **Core Kubernetes API kinds:** Simple configuration and monitoring related types such as `ConfigMap`, `Secrets` and `Events`.
+There should be no need for a stack to create a Deployment, Pod, Volume, or other types like that.
+* **Owned CRDs:** All CRDs that the stack owns and are being installed by the stack.
+* **Dependent CRDs:** All CRDs that the stack directly depends on and will consume to perform its primary operations (e.g. `MySQLInstance` claim).
```

Comment:
I assume the `verbs` granted will be different for owned vs dependent crds.

Reply:
Same question - this seems like an area where someone may want finer-grained control over exactly which permissions the stack has

Reply:
Actually, I'm not sure that we should make a distinction in the verbs granted for owned vs. dependent CRDs.  Consider the case of the WordPress stack that depends on the MySQLInstance CRD.  Off the top of my head, The WordPress controller will need to create, update, and delete instances of the MySQLInstance CRD.  I could see it also wanting get and watch as well.  What's left after that, list?

I can see very valid reasons for a stack to do most any verb on both owned and dependent CRDs, so I'm thinking we shouldn't make a distinction right now.

Reply:
This isn't a discussion that I feel needs to block merging, but an example would be if I am authoring a stack which inspects resources managed by other stacks. Maybe it's a stack for analytics or instrumentation. Or a stack that wants read-only access for some other reason, like replicating, backing up, or posting things to Slack for some reason. As the author of that stack, I wouldn't want to accidentally delete the things I was trying to watch, and as a security-minded person who installs that stack, I wouldn't want a stack that I expected to only need read permissions to also have write permissions.

Reply:
That's a valid point, thank you for pushing on this.  I'm wondering if we should try to address this fairly loose access to verbs on CRDs that a stack depends on still within this milestone.

One potential approach could be when the stack declares the CRDs that it depends on, it could optionally declare the verbs it needs as well.  By default it will get all verbs, but if the stack wants to scope down the verbs, maybe it should be able to?

This could potentially be a fair trade-off between scoping permissions down to the minimal set, not allowing access to built-in types such as Pods, and being manageable to implement. Thoughts?

Example:
```
dependsOn:
- kind: Foo
  apiversion: mygroup.io/v1alpha1
  verbs: "get,create,update,delete"
- kind: Bar
  apiversion: yourgroup.io/v1alpha2
  verbs: "get,watch"
```

Reply:
This idea needs to be thought through a bit more, so I've captured it in #758,